### PR TITLE
Erc20 complete

### PR DESCRIPTION
**Pull request by lexfrl**
Complete impl of ERC-20 token:
- 2 transactions:
fn transferFrom(&mut self, _from: Address, _to: Address, _amount: U256) -> bool;
fn approve(&mut self, _spender: Address, _value: U256) -> bool;
- 1 request:
fn allowance(&mut self, _owner: Address, _spender: Address) -> U256;
- 1 event:
fn Approval(&mut self, indexed_owner: Address, indexed_spender: Address, _value: U256);

### CODE REVIEW

File: src/token.rs [hunk 1]
```diff
@@ -229,4 +299,78 @@ mod tests {
...
+            assert_eq!(contract.balanceOf(samAddress.clone()), 10000.into());
+            assert_eq!(contract.balanceOf(owner.clone()), 30000.into());
+        }
+    );
+
+    test_with_external!(
+        ExternalBuilder::new()
+            // Allowance to spend from tokens owner for spender
+            .storage("e428ed52e2e5765d58bbc65412f7c432d6388f27105bbb1ede5c2badfc4b3b6b".into(), U256::from(50000).into())
+            // Total Sypply
```
Comment (pepyakin) [inline src/token.rs]: typo: supply

File: src/token.rs [hunk 2]
```diff
@@ -229,4 +299,78 @@ mod tests {
...
+            assert_eq!(contract.transferFrom(owner.clone(), samAddress.clone(), 1.into()), false);
+            assert_eq!(contract.balanceOf(samAddress.clone()), 10000.into());
+            assert_eq!(contract.balanceOf(owner.clone()), 30000.into());
+        }
+    );
+
+    test_with_external!(
+        ExternalBuilder::new()
+            // Allowance to spend from tokens owner for spender
+            .storage("e428ed52e2e5765d58bbc65412f7c432d6388f27105bbb1ede5c2badfc4b3b6b".into(), U256::from(50000).into())
```
Comment (NikVolf) [inline src/token.rs]: why do you do setup with storage directly, not with contract methods?

File: src/token.rs [hunk 3]
```diff
@@ -229,4 +299,78 @@ mod tests {
...
+                "0x0000000000000000000000000000000000000000000000000000000000000000".into(), // sender (owner) address
+                "0x000000000000000000000000db6fd484cfa46eeeb73c71edee823e4812f9e2e1".into()]); // spender address
+            assert_eq!(contract.allowance(::pwasm_std::ext::sender(), spender.clone()), 40000.into());
+        }
+    );
+
+    test_with_external!(
+        ExternalBuilder::new()
+            // Allowance to spend from tokens owner for spender
+            .storage("e428ed52e2e5765d58bbc65412f7c432d6388f27105bbb1ede5c2badfc4b3b6b".into(), U256::from(10000).into())
```
Comment (NikVolf) [inline src/token.rs]: ditto

File: Cargo.lock [hunk 4]
```diff
@@ -143,7 +196,7 @@ dependencies = [
 
 [[package]]
 name = "tiny-keccak"
-version = "1.3.1"
```
Comment (NikVolf) [inline Cargo.lock:146]: is it the same version as in pwasm-std?
better use the same

File: Cargo.toml [hunk 5]
```diff
@@ -7,6 +7,8 @@ authors = ["Alexey Frolov <frol.rage@gmail.com>"]
 pwasm-std = "0.2"
 pwasm-abi = { git = "https://github.com/paritytech/pwasm-abi" }
 pwasm-abi-derive = { git = "https://github.com/paritytech/pwasm-abi", path = "derive" }
+tiny-keccak = "1.3"
```
Comment (pepyakin) [inline Cargo.toml:10]: I wondering, how bigger binary size will get with this dependency.

File: src/token.rs [hunk 6]
```diff
@@ -45,20 +47,64 @@ pub mod contract {
...
+    fn read_balance_of(owner: &Address) -> U256 {
+        storage::read(&balance_key(owner)).into()
+    }
+
+    // Reites allowance value using a key, generated by allowance_key
+    fn read_allowance(key: &H256) -> U256 {
+        storage::read(key).into()
+    }
+
+    // Writes allowance value using a key, generated by allowance_key
```
Comment (pepyakin) [inline src/token.rs]: I think you mean that 

> it is expected that key will be generated by `allowance_key`

File: src/token.rs [hunk 7]
```diff
@@ -85,26 +131,50 @@ pub mod contract {
...
+            let recipientBalance = read_balance_of(&to);
             if amount == 0.into() || senderBalance < amount {
                 false
             } else {
-                senderBalance = senderBalance - amount;
-                recipientBalance = recipientBalance + amount;
-                storage::write(&balance_key(&sender), &senderBalance.into());
-                storage::write(&balance_key(&to), &recipientBalance.into());
+                storage::write(&balance_key(&sender), &(senderBalance - amount).into());
+                storage::write(&balance_key(&to), &(recipientBalance + amount).into());
```
Comment (pepyakin) [inline src/token.rs]: Would it be better to extract expressions like `recipientBalance + amount` into thier own statements. I personally think that these operations are pretty important to be maximally visible for readers of the code.

Comment (pepyakin) [reply src/token.rs]: (here and below)

File: src/token.rs [hunk 8]
```diff
@@ -229,4 +299,78 @@ mod tests {
...
+            .storage("0100000000000000000000000000000000000000000000000000000000000000".into(), U256::from(40000).into())
+            .sender("0xdb6fd484cfa46eeeb73c71edee823e4812f9e2e1".into())
+            .build(),
+        spender_should_not_be_able_to_spend_if_no_owner_has_no_coins {
+            let mut contract = TokenContractInstance{};
+            contract.constructor(40000.into());
+            let owner: Address = Address::new();
+            let samAddress: Address = "0xea674fdde714fd979de3edf0f56aa9716b898ec8".into();
+
+            // Cant transfer because owner has no
```
Comment (pepyakin) [inline src/token.rs]: "has no" what?

File: src/token.rs [hunk 9]
```diff
@@ -229,4 +299,78 @@ mod tests {
...
+
+            assert_eq!(contract.transferFrom(owner.clone(), samAddress.clone(), 5000.into()), true);
+            assert_eq!(contract.balanceOf(samAddress.clone()), 5000.into());
+            assert_eq!(contract.balanceOf(owner.clone()), 35000.into());
+
+            assert_eq!(contract.transferFrom(owner.clone(), samAddress.clone(), 5000.into()), true);
+            assert_eq!(contract.balanceOf(samAddress.clone()), 10000.into());
+            assert_eq!(contract.balanceOf(owner.clone()), 30000.into());
+
+            // limit reached. No more coins should be available to spent for the spender
```
Comment (pepyakin) [inline src/token.rs]: I'm bad at english, but isn't this should be "for spending" instead of "to spent"?

File: Cargo.toml (hunk 5 above)
Comment (pepyakin) [reply Cargo.toml:10]: Ah, I see, it is already being used by `pwasm-std`.

File: src/token.rs (hunk 9 above)
Comment (NikVolf) [reply src/token.rs]: or at least "to spend"

**Comment by NikVolf** [comment]
Also please copy json from `./target/json` to `./compiled` in build script

**Pull request merged**